services/gateway/shared/registerHttpApi.js

```diff
@@ -64,137 +94,135 @@ const configureApi = (apiNames, apiPrefix, registeredModuleNames) => {
+	},
+});
 
-			const unknownList = Object.keys(paramReport.unknown);
-			if (unknownList.length > 0) {
-				sendResponse(INVALID_REQUEST[0], `Unknown input parameter(s): ${unknownList.join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			if (paramReport.required.length) {
-				sendResponse(INVALID_REQUEST[0], `Require one of the following parameter combination(s): ${paramReport.required.join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			const invalidList = paramReport.invalid;
-			if (invalidList.length > 0) {
-				sendResponse(INVALID_REQUEST[0], `Invalid input: ${invalidList.map(o => o.message).join('; ')}`);
-				throw new ValidationException('Request param validation error.');
-			}
-
-			const params = transformRequest(methodPaths[routeAlias], dropEmptyProps(paramReport.valid));
-			req.$params = params;
-		},
-
-		async onAfterCall(ctx, route, req, res, data) {
-			if (gatewayConfig.api.enableHttpCacheControl) {
-				// Set 'Cache-Control' to assist response caching
-				ctx.meta.$responseHeaders = { 'Cache-Control': gatewayConfig.api.httpCacheControlDirectives };
-			}
-
-			// Set response headers and return CSV data if filename available
-			if (data.data && data.meta && data.meta.filename && data.meta.filename.endsWith('.csv')) {
-				res.setHeader('Content-Disposition', `attachment; filename="${data.meta.filename}"`);
-				res.setHeader('Content-Type', 'text/csv');
-				res.end(data.data);
-				return res;
-			}
-
-			if (data.data && data.status) {
-				if (data.status === 'SERVICE_UNAVAILABLE') ctx.meta.$responseHeaders = { 'Retry-After': 30 };
-				if (data.status === 'INVALID_PARAMS') data.status = 'BAD_REQUEST';
-
-				ctx.meta.$statusCode = StatusCodes[data.status] || data.status;
-
-				// 'ACCEPTED' is a successful HTTP code
-				if (data.status !== 'ACCEPTED') {
-					let message = `The request ended up with ${data.status} error`;
-
-					if (typeof data.data === 'object' && typeof data.data.error === 'string') {
-						message = data.data.error;
-					}
-
-					return {
-						error: true,
-						message,
-					};
-				}
-			}
-			const apiPath = `${req.method.toUpperCase()} ${req.$alias.path}`;
-			return transformResponse(methodPaths[apiPath], data);
-		},
-	};
+const registerApi = (apiNames, config, registeredModuleNames) => {
+	const allAPIs = getAllAPIs(apiNames, registeredModuleNames);
+	const apisToRegister = [];
+
+	const ALLOWED_ETAG_VALUES = ['strong', 'weak', true, false, 'custom'];
+	const DEFAULT_ETAG_VALUE = 'strong';
+
+	// eslint-disable-next-line no-restricted-syntax
+	for (const eTagVal of ALLOWED_ETAG_VALUES) {
+		const etagAPIs = Object.fromEntries(Object.entries(allAPIs)
+			.filter(([, value]) => {
+				if (value.etag === undefined) value.etag = DEFAULT_ETAG_VALUE;
+				return (eTagVal === 'custom' && typeof value.etag === 'function') || value.etag === eTagVal;
+			}));
```

Comment:
This can be simplified like below where we do not need the outer loop:

```suggestion
Object.entries(allAPIs).forEach(([key, apiConf]) => {
	const eTag = apiConf.etag || DEFAULT_ETAG_VALUE;

	if (ALLOWED_ETAG_VALUES.includes(eTag)) {
		apisToRegister.push({
			[key]: {
				...apiConf,
				tag: eTag,
			},
		});
	}
});
```

Reply:
Also, since we are using `ALLOWED_ETAG_VALUES` only here, we can make it a `Set` for faster lookup. But this is quite optional as this script will only be invoked on service start.